api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto

```diff
@@ -891,6 +891,9 @@ message HttpConnectionManager {
   // This should be set to ``false`` in cases where Envoy's view of the downstream address may not correspond to the
   // actual client address, for example, if there's another proxy in front of the Envoy.
   google.protobuf.BoolValue add_proxy_protocol_connection_state = 53;
+
+  // If true, envoy will retain keep-alive response headers from upstream for http1.1
```

Comment:
I think this needs to be a much longer explanation. Specifically, it should clear say which RFC this violates, why it violates it, and when someone would want to set this.

Reply:
Also, this change keeps the request portion of the change also, so if we are keeping that, this name should change also, probably to `retain_keepalive_headers`. The longer comment should also clearly indicate what happens if either half is HTTP/2.

Reply:
If setting specific connection headers would be sufficient for this user's use case I'd  much rather go that route.  @noah8713 would set keep-alive work for you or do you actually need preservation of different headers from different upstreams?

Reply:
See also https://github.com/envoyproxy/envoy/issues/3011 and https://github.com/envoyproxy/envoy/issues/17613.

I would be in favor of giving users control of connection headers if they wish per these linked issues which I think will also fix your case per @alyssawilk

Reply:
+1 - if it could be in response_headers_to_add and be cleared for non-HTTP/1.1 that'd be fantastic.

Reply:
> I think this needs to be a much longer explanation. Specifically, it should clear say which RFC this violates, why it violates it, and when someone would want to set this.

Thanks yes I had it in commit details and should have extended in comments too. Will do part of amendments

Reply:
> +1 - if it could be in response_headers_to_add and be cleared for non-HTTP/1.1 that'd be fantastic.

Thanks @mattklein123 @alyssawilk seems this seems to be good path forward and looks neat. Saves us api change. So I will amend the logic to validate 2 things

1. if protocol h11 and request_headers has connection:keep-alive, retain connection-keep-alive
2. if protocol h11 and response_headers has keep-alive: timeout=xx, retain keep-alive response header. 


Will rectify the change logs accordingly part of amendments.

Reply:
I don't think that's what we're suggesting.   Always doing pass-through would be a very negative change for most Envoy operatators.
Envoy has a mechanism to add headers: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route.proto request_headers_to_add / response_headers_to_add.
I'm suggesting that if what you need is to send a keep-alive header, _and_ if configuring it on a per-route or a per-hcm basis is sufficient, we should use that existing mechanism (iff the headers are being sent over an HTTP 1.1 codec)

If you absolutely need to preserve a timeout that Envoy will not actually enforce, you will need a different mechanism.  In that case I'd suggest  general 'do not clear hop by hop headers' config option, and you can use a filter to clear the ones you want.  I really would like to avoid having a knob for standards-breaking preservation of individual hop by hop headers.

Reply:
Thanks.  I dont think its  a new knob. I was referring to existing request_headers as request_headers_to_add which is already exposed (my bad should have explicitly stated the actual vars request_headers_to_add and response_headers_to_add ). Adding any custom headers are all good as long as envoy is not discarding in code as is unlike for connection and keep-alive. Just leveraging request_headers_to_add / response_headers_to_add as istio too provides changing headers via [virtual services](https://istio.io/latest/docs/reference/config/networking/virtual-service/#Headers) which underlying uses https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route.proto  or we can use it for entire mesh with envoyfilter to apply these settings for all (thats upto users and operators to control). What I think is we still need to ensure its present in headers if users dont explicitly set it. However,  that too will not work in current case as its being discarded in code. So  will just enhance the logic to see if response_header_to_add has keep-alive and its h11 protocol keep it; Discard for other protocols. Also `Connection:keep-alive` is for request not for response and clear hop by hop only clears connection header in response not request. .Because its not just clear hop-by-hop knob, also  https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_manager_utility.cc#L102  and https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_manager_utility.cc#L93. needs to be amended.  Let me know if I misunderstood from code side as also recall you have already called it out https://github.com/envoyproxy/envoy/issues/17613#issuecomment-896103616 need to filter between the ones set by user with request/response_headers_to_add vs from codec. Its just response_header_to_add is per listener vs global change. That is why adding via config knob or runtime guard was also valid one to avoid mass refactor. Let me know else I am ok to keep the HCM config too to avoid over complicating/over engineering

Reply:
> If setting specific connection headers would be sufficient for this user's use case I'd much rather go that route. @noah8713 would set keep-alive work for you or do you actually need preservation of different headers from different upstreams?

Other custom headers are ok. Just need to retain these two connection:keep-alive in request and keep-alive in response as its hardcoded in code without protocol check or any knob control.


e.g. 
```
"response_headers_to_add": [
         {
          "header": {
           "key": "keep-alive",
           "value": "timeout=60"
          }
         },
         {
          "header": {
           "key": "x-test-me",
           "value": "ok-fine"
          }
         }
```

but because its hard coded in code, it gets discarded 
```
 curl --http1.1 -s -I localhost:10000
HTTP/1.1 200 OK
xxxx
x-test-me: ok-fine
```

Main reason is to make it easier for wider community to leverage the same and also saves us to maintain it in house for each envoy upgrade. That is why kept it per HCM config or runtime guard. Using HCM was beneficial to have it configure per listener. Amending via response_header_to_add field will need to iterate router config and it seems for one header need to do too much refactoring :) . Anyways I am ok to sync up during next meetup too and discuss but this is what we have in hand with all option. Please confirm accordingly @mattklein123 @alyssawilk :)

Reply:
I think what @alyssawilk is proposing which I agree with is to make request/response headers to add respect the connection headers and not remove them. If this works for your use case I think that is optimal.

Reply:
Thanks @mattklein123 I did spend some time reading and tryout few bits but I am not clear from code at-least as clear_hop_by_hop removes it anyways ref https://github.com/envoyproxy/envoy/pull/17110 its only allowed for envoy mobile to passthrough list of these headers; verified multiple times it doesn't work as is even on setting it on req/response headers to add.  Our goal here is same to  passthrough the response `keep-alive:timeout=xx` settings from server/upstream as is to downstream so that connection can be reused as different apps have different keep-alive:timeout settings. Since clear_hop_by_hop is not exposed for HCM to turn it off or even over-ride;  per suggestion from @alyssawilk ask is to not change hop by hop logic for this case.  That is why exposed this change via config knob to return as is via first cut. Can we keep the config knob via hcm for time being? Later on we can amend setting for each application behind proxy via response_header_to_add and amend the clear hop by hop logic if set. Basically upstream/server can set that part of onboarding their service behind proxy and envoy can keep honoring it for http1.1. Not sure if it can be done without any major code changes but open to suggestions to move forward.